lib/spack/spack/build_environment.py

```diff
@@ -1437,3 +1424,51 @@ def write_log_summary(out, log_type, log, last=None):
         # If no errors are found but warnings are, display warnings
         out.write("\n%s found in %s log:\n" % (plural(nwar, "warning"), log_type))
         out.write(make_log_context(warnings))
+
+
+class ModuleMonkeyPatcher:
```

Comment:
What do you think of a name like `ModuleChangeAccumulator` instead? I don't think what we're doing is properly monkeypatching here.

Reply:
You're right it's not proper monkey-patching. I went with `ModuleChangePropagator`. I'll take care of Python 2 compatibility when backporting this to the release branch (mainly f-strings and not deriving from object).